### PR TITLE
Handle nested fields_for by adding indexes to record_name

### LINKED ISSUE #15332: Wrong form field names in nested fields_for
I have a view with structure like this:
```
<%= form_for :foos, url: root_path do |f| %>
  <% @foos.each do |foo| %>
    <%= f.fields_for "foo[]", foo do |f2| %>
      <%= f2.fields_for :bars do |b| %>
        <%= f2.text_field :id %>
        <%= b.text_field :id %>
      <% end %>
    <% end %>
  <% end %>
<% end %>
```
and there is an issue with inner form fields naming: while outer(f2) fields has names that starts with 'foos[foo][]', inner(b) fields has names like 'foos[foo[]]', so submitting of such form fails with something like 'expected Array (got Rack::Utils::KeySpaceConstrainedParams) for param `foo''.

This issue is reproducible both in 4.1.1 and master, test repo can be found at https://github.com/ojab/nested_fail

### PR DESCRIPTION
**Pull request by ojab**
Fixes #15332

### CODE REVIEW

**Comment by sgrif** [comment]
Can you please update your commit message to better express what the issue this is fixing is, why this fix works, and what the use case is? Even with the issue linked I'm having trouble following, and we like to keep as much context in the commits as possible.

**Comment by ojab** [comment]
Updated commit message, hope that the issue subject is more clear now.

**Comment by rafaelfranca** [comment]
I still think this is not the correct usage. With only one nested attribute the name format is: `post[comments_attributes][0][body]` I'd expect with who the format to be `post[post_attributes][0][comments_attributes][0][body]`

**Comment by rafaelfranca** [comment]
Yeah, it should be the same as https://github.com/ojab/rails/blob/master/actionview/test/template/form_helper_test.rb#L2326

**Comment by ojab** [comment]
Both variants (foos[foo[0]][bar[1]][id] & foos[foo][0][bar][1]][id]) producing identical params hash, but yeah, second is more readable.

Pushed fixed version.

File: actionview/lib/action_view/helpers/form_helper.rb [hunk 1]
```diff
@@ -1604,6 +1604,9 @@ def fields_for(record_name, record_object = nil, fields_options = {}, &block)
         when String, Symbol
           if nested_attributes_association?(record_name)
             return fields_for_with_nested_attributes(record_name, record_object, fields_options, block)
+          elsif !defined?(@auto_index) && !options[:index] && record_name.to_s.end_with?('[]')
+            record_name = record_name.to_s.sub(/(.*)\[\]$/, "[\\1][#{record_object.id}]")
```
Comment (egilburg) [inline actionview/lib/action_view/helpers/form_helper.rb]: perhaps do this line in the blow `if nested` block to keep all string building in one place

Comment (ojab) [inline actionview/lib/action_view/helpers/form_helper.rb]: Actually the whole part is unneeded, pushed another one version.

**Comment by ojab** [comment]
ping?

**Comment by ojab** [comment]
Also I've run a benchmark for the `else` case, i. e.
```ruby
      form_for(:posts) do |f|
        f.fields_for('post[]', @post) do |f2|
          f2.text_field(:id)
        end
      end
```
but results are inconclusive, I'm getting very close results with large stddev:
```
Calculating -------------------------------------
                orig   235.000  i/100ms
                gsub   247.000  i/100ms
              substr   251.000  i/100ms
-------------------------------------------------
                orig      2.530k (±12.7%) i/s -     49.820k
                gsub      2.664k (±13.4%) i/s -     50.882k
              substr      2.609k (±12.2%) i/s -     49.196k
```
orig -- rails without patch, gsub -- rails with this PR applied, substr -- record_name is formed using `"#{object_name}[#{record_name.to_s[0..-3]}][#{record_object.id}]"` without gsub.
So AFAIU performance change is negligible and gsub is as good as substring.

**Comment by paratiger** [comment]
Hi,

I've come across a situation that is effected by this change and wanted to check that this is correct

I use `fields_for` to add non active record objects that are stored in a serialized field, for example

```
class Page < ActiveRecord::Base
  serialize :content_blocks, Array
end
```

```
class ContentBlock
...
end
```

Now using `f.fields_for "content_blocks[]", ContentBlock.new do |ff|` fails as it doesn't respond to `id`

I can get around this by passing an empty index `index: ''` to give me `page[content_blocks][][title]` but should the method rescue when the `record_object` doesn't respond to `id`

**Comment by ojab** [comment]
I'm not quite sure what exactly you want to archive. [My testcase (using rails-4.2.6)](https://github.com/ojab/test21431/commit/4f0ef85b7e58b9701a61ce43facf49a0abc30d3f) produces input field w/ `ContentBlock.new.to_s` as id, i. e. something like:
```
<input value="some content" name="page[content_blocks][#<ContentBlock:0x007fbcf6bfc790>][block_content]" id="page_content_blocks_#<ContentBlock:0x007fbcf6bfc790>_block_content" type="text">
```
and I doubt that it is what you want.
Can you upload a testcase that reproduces your issue?

**Comment by rocketblr** [comment]
Hi @ojab 
To be honest I'm not sure if it was a bug or a feature, but this pull request broke the possibility of using fields_for with arrays, because they don't respond to `.id` method. We use the following approach in our project:
```
<%= form_for :bulk do |t| %>
  <%= t.fields_for 'people[]', [] do |p| %>
    First Name: <%= p.text_field :first_name %>
    Last Name: <%= p.text_field :last_name %>
  <% end %>
<% end %>
```
And params looks like:
```
"bulk" => {
  "people" => [
    {"first_name" => "Michael", "last_name" => "Jordan"},
    {"first_name" => "Steve", "last_name" => "Jobs"},
    {"first_name" => "Barack", "last_name" => "Obama"}
  ]
}
```
It works perfect in Rails 4.2.5, but fails in Rails 5.

Again I don't know if this behavior of fields_for was a bug or a feature, but it looks like there is no way to get params as array using fields_for now.

**Comment by ojab** [comment]
As far as I understand, this form is supposed to create a form w/ fields for one Person and other people's fields are created via AJAX or something similar, isn't it?
If so, I believe that you can use `t.fields_for :people, index: nil` instead.

**Comment by rocketblr** [comment]
I swear I've tried that. My bad. Working as expected. Thanks! Sorry!

**Pull request merged**